cmd/cli/update.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+	helm "helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/strvals"
+
+	"github.com/openservicemesh/osm/pkg/cli"
+)
+
+const updateDesc = `
+This command updates an osm control plane on the Kubernetes cluster.
+
+The default Kubernetes namespace that gets updated is called osm-system. 
+To update a control plane components in a different namespace, use the 
+--namespace flag.
+
+Example:
+  $ osm update --namespace hello-world --enable-egress false
+`
+
+type updateCmd struct {
+	out io.Writer
+
+	chartPath                     string
+	enablePermissiveTrafficPolicy bool
+	enableEgress                  bool
+	meshCIDRRanges                []string
+}
+
+func newUpdateCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
+	upd := &updateCmd{
+		out: out,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "update osm control plane configuration",
+		Long:  updateDesc,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return upd.run(config)
+		},
+	}
+
+	f := cmd.Flags()
+
+	f.StringVar(&upd.chartPath, "osm-chart-path", "", "path to osm chart to override default chart")
+	f.BoolVar(&upd.enablePermissiveTrafficPolicy, "enable-permissive-traffic-policy", false, "Enable permissive traffic policy mode")
+	f.BoolVar(&upd.enableEgress, "enable-egress", false, "Enable egress in the mesh")
+	f.StringSliceVar(&upd.meshCIDRRanges, "mesh-cidr", []string{}, "mesh CIDR range, accepts multiple CIDRs, required if enable-egress option is true")
+
+	return cmd
+}
+
+func (u *updateCmd) run(config *helm.Configuration) error {
+	var chartRequested *chart.Chart
```

Comment:
This is a general question. Not a comment. Generally speaking you want to have run sequenced as:
1. default input if needed
2. validate input as needed
3. run..

This even works better if and when you choose to use versioned options/flags.

Do you know why aren't we doing that?

Reply:
Do you mean something like this, Kal? https://github.com/openservicemesh/osm/blob/b7337ab144d0126fa44312bc3001f563183ffb76/cmd/cli/install.go#L199

We use the model for install where there are many arguments that are passed in. We could do the same to validate if we can load the chart